bin/faucet/src/faucet/mod.rs

```diff
@@ -0,0 +1,422 @@
+    pub note_type: NoteType,
+    /// The amount to mint.
+    pub asset_amount: AssetAmount,
+}
+
+type MintResult<T> = Result<T, MintError>;
+
+/// Error indicating what went wrong in the minting process for a request.
+#[derive(Debug, thiserror::Error)]
+pub enum MintError {
+    #[error("building the p2id note failed")]
+    BuildingP2IdNote(#[source] NoteError),
+    #[error("compiling the tx script failed")]
+    ScriptCompilation(#[source] TransactionScriptError),
+    #[error("execution of the tx script failed")]
+    Execution(#[source] TransactionExecutorError),
+    #[error("proving the tx failed")]
+    Proving(#[source] TransactionProverError),
+    #[error("submitting the tx to the node failed")]
+    Submission(#[source] RpcError),
+}
+
+/// Stores the current faucet state and handles minting requests.
+pub struct Faucet {
+    authenticator: Arc<BasicAuthenticator<StdRng>>,
+    data_store: Arc<FaucetDataStore>,
+    id: FaucetId,
+    // Previous faucet account states used to perform rollbacks if a desync is detected.
+    prior_state: VecDeque<Account>,
+}
+
+impl Faucet {
+    /// Loads the faucet state from the node and the account file.
+    #[instrument(name = "Faucet::load", fields(id), skip_all)]
+    pub async fn load(
+        account_file: AccountFile,
+        rpc_client: &mut RpcClient,
+    ) -> anyhow::Result<Self> {
+        let id = account_file.account.id();
+        let id = FaucetId(id);
+
+        tracing::Span::current().record("id", id.to_string());
+
+        info!("Fetching faucet state from node");
+
+        let account = match rpc_client.get_faucet_account(id).await {
+            Ok(account) => {
+                info!(
+                    commitment = %account.commitment(),
+                    nonce = %account.nonce(),
+                    "Received faucet account state from the node",
+                );
+
+                Ok(account)
+            },
+            Err(RpcError::Transport(status)) if status.code() == Code::NotFound => {
+                let account = account_file.account;
+                info!(
+                    commitment = %account.commitment(),
+                    nonce = %account.nonce(),
+                    "Faucet not found in the node, using state from file"
+                );
+
+                Ok(account)
+            },
+            Err(err) => Err(err),
+        }
+        .context("fetching faucet state from node")?;
+
+        info!("Fetching genesis header from the node");
+        let genesis_header = rpc_client
+            .get_genesis_header()
+            .await
+            .context("fetching genesis header from the node")?;
+
+        // SAFETY: An empty chain MMR should be valid.
+        let genesis_chain_mmr = ChainMmr::new(
+            PartialMmr::from_peaks(
+                MmrPeaks::new(0, Vec::new()).expect("Empty MmrPeak should be valid"),
+            ),
+            Vec::new(),
+        )
+        .expect("Empty ChainMmr should be valid");
```

Comment:
Unrelated to this PR - should those crypto types provide a default impl or zero() -like constructors so we don't have to unwrap/expect in these situations?

Reply:
That would be lovely - @PhilippGackstatter any chance of an empty/default maybe?

Reply:
I'm happy to make this change. Keep me sane while I do the terraform stuff 😃

Reply:
https://github.com/0xMiden/crypto/pull/409

Then need to update base for `ChainMmr` I suppose.

Reply:
Thanks!

We need to be careful about changing base (or any further deps). This can cause issues with the client if we pull in any unforseen changes.

I'd rather do it as part of trunk for the next release.